File: backend/app/schemas/lead.py
"""
Lead schemas for request/response validation.
"""
from typing import Optional, List, Any, Dict
from pydantic import BaseModel, EmailStr, Field


class LeadBase(BaseModel):
    """Base lead schema with common fields."""
    
    First_Name: Optional[str] = Field(None, description="Lead's first name")
    Last_Name: str = Field(..., description="Lead's last name (required)")
    Email: Optional[EmailStr] = Field(None, description="Lead's email address")
    Phone: Optional[str] = Field(None, description="Lead's phone number")
    Mobile: Optional[str] = Field(None, description="Lead's mobile number")
    Company: Optional[str] = Field(None, description="Lead's company name")
    Title: Optional[str] = Field(None, description="Lead's job title")
    Industry: Optional[str] = Field(None, description="Industry")
    Lead_Source: Optional[str] = Field(None, description="How the lead was acquired")
    Lead_Status: Optional[str] = Field(None, description="Current lead status")
    Website: Optional[str] = Field(None, description="Website URL")
    Description: Optional[str] = Field(None, description="Additional notes")
    
    # Address fields
    Street: Optional[str] = None
    City: Optional[str] = None
    State: Optional[str] = None
    Zip_Code: Optional[str] = None
    Country: Optional[str] = None


class LeadCreate(LeadBase):
    """Schema for creating a new lead."""
    pass


class LeadUpdate(BaseModel):
    """Schema for updating a lead (all fields optional)."""
    
    First_Name: Optional[str] = None
    Last_Name: Optional[str] = None
    Email: Optional[EmailStr] = None
    Phone: Optional[str] = None
    Mobile: Optional[str] = None
    Company: Optional[str] = None
    Title: Optional[str] = None
    Industry: Optional[str] = None
    Lead_Source: Optional[str] = None
    Lead_Status: Optional[str] = None
    Website: Optional[str] = None
    Description: Optional[str] = None
    Street: Optional[str] = None
    City: Optional[str] = None
    State: Optional[str] = None
    Zip_Code: Optional[str] = None
    Country: Optional[str] = None


class LeadResponse(BaseModel):
    """Response schema for a single lead."""
    
    data: Dict[str, Any] = Field(..., description="Lead data from Zoho")
    
    class Config:
        extra = "allow"


class LeadListResponse(BaseModel):
    """Response schema for a list of leads."""
    
    data: List[Dict[str, Any]] = Field(default_factory=list, description="List of leads")
    page: int = Field(1, description="Current page number")
    per_page: int = Field(50, description="Records per page")
    total_count: int = Field(0, description="Total number of records in this response")
    more_records: bool = Field(False, description="Whether more records are available")

File: frontend/src/types/lead.ts
// AI-generated analysis for a lead
export interface LeadAnalysis {
  company_name?: string;
  country?: string;
  region?: string;
  summary?: string;
  product_description?: string;
  vertical?: string;
  business_model?: string;
  motion?: string;
  raise_stage?: string;
  company_size?: string;
  likely_icp_canada?: string;
  fit_score?: number;
  fit_assessment?: string;
  key_insights?: string[];
  questions_to_ask?: string[];
  confidence_level?: string;
  notes?: string[];
}

// Marketing material matched by semantic similarity
export interface MarketingMaterial {
  material_id: string;
  title: string;
  link: string;
  industry?: string;
  business_topics?: string;
  similarity_score?: number;
}

// Similar customer identified by LLM analysis
export interface SimilarCustomer {
  name: string;
  description: string;
  industry?: string;
  website?: string;
  why_similar?: string;
}

export interface Lead {
  id: string;
  First_Name?: string;
  Last_Name: string;
  Email?: string;
  Phone?: string;
  Mobile?: string;
  Company?: string;
  Title?: string;
  Industry?: string;
  Lead_Source?: string;
  Lead_Status?: string;
  Website?: string;
  Description?: string;
  Street?: string;
  City?: string;
  State?: string;
  Zip_Code?: string;
  Country?: string;
  Created_Time?: string;
  Modified_Time?: string;
  Owner?: {
    id: string;
    name: string;
    email: string;
  };
  // Additional custom fields for your dashboard
  Raise?: string;
  Verticle?: string;
  Business_Model?: string;
  Fit_Tag?: string;
  Fit_Relationship?: string;
  Typical_Customer?: string;
  Target_Group?: string;
  Questions_To_Ask?: string;
  Marketing_Material?: string;
  // Allow for any additional Zoho fields
  [key: string]: unknown;
}

// Lead with analysis data (used when fetching single lead details)
export interface LeadWithAnalysis {
  data: Lead;
  analysis?: LeadAnalysis;
  analysis_available?: boolean;
  from_cache?: boolean;
  marketing_materials?: MarketingMaterial[];
  similar_customers?: SimilarCustomer[];
}

export interface LeadListResponse {
  data: Lead[];
  page: number;
  per_page: number;
  total_count: number;
  more_records: boolean;
}

export interface LeadResponse {
  data: Lead;
  analysis?: LeadAnalysis;
  analysis_available?: boolean;
  from_cache?: boolean;
  marketing_materials?: MarketingMaterial[];
  similar_customers?: SimilarCustomer[];
}
